source/particle/generator/uniform_box.cc

```diff
@@ -0,0 +1,187 @@
+/*
+  Copyright (C) 2015 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file doc/COPYING.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/generator/uniform_box.h>
+
+#include <boost/random.hpp>
+#include <deal.II/base/std_cxx11/array.h>
+
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Generator
+    {
+      template <int dim>
+      UniformBox<dim>::UniformBox() {}
+
+      template <int dim>
+      std::multimap<types::LevelInd, Particle<dim> >
+      UniformBox<dim>::generate_particles()
+      {
+        std::multimap<types::LevelInd, Particle<dim> > particles;
+
+        types::particle_index cur_id = 0;
+        const Tensor<1,dim> P_diff = P_max - P_min;
+
+        double volume(1.0);
+        for (unsigned int i = 0; i < dim; ++i)
+          volume *= P_diff[i];
+
+        std_cxx11::array<double,dim> nParticles;
+        std_cxx11::array<double,dim> spacing;
+
+        // Calculate separation of particles
+        for (unsigned int i = 0; i < dim; ++i)
+          {
+            nParticles[i] = round(std::pow(n_tracers * std::pow(P_diff[i],dim) / volume, 1./dim));
+            spacing[i] = P_diff[i] / fmax(nParticles[i] - 1,1);
+          }
+
+        for (types::particle_index i = 0; i < nParticles[0]; ++i)
+          {
+            for (types::particle_index j = 0; j < nParticles[1]; ++j)
+              {
+                if (dim == 2)
+                  {
+                    const Point<dim> particle_position = Point<dim> (P_min[0]+i*spacing[0],P_min[1]+j*spacing[1]);
+                    particles.insert(this->generate_particle(particle_position,cur_id++));
```

Comment:
I often write this as
```
  const Point<dim> particle_position = (dim == 2
                                                              ?
                                                              Point<dim>(x,y)
                                                              :
                                                              Point<dim>(x,y,z));
```
This way you avoid the `if` statement and code duplication. It throws an assertion if `dim` is neither 2 or 3 because you end up in the second constructor which is only valid in 3d.